source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -829,38 +823,62 @@ AdaptiveNamespaceStart
                 XamlHelpers::SetContent(localButton.Get(), title.Get());
                 return;
             }
-            XamlHelpers::AppendXamlElementToPanel(buttonIcon.Get(), buttonContentsPanel.Get()); // Add image to stack panel
+
+            // Create title text block
+            ComPtr<ITextBlock> buttonText = XamlHelpers::CreateXamlClass<ITextBlock>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_TextBlock));
+            THROW_IF_FAILED(buttonText->put_Text(title.Get()));
+            THROW_IF_FAILED(buttonText->put_TextAlignment(TextAlignment::TextAlignment_Center));
+
+            // Handle different arrangements inside button
             ComPtr<IFrameworkElement> buttonIconAsFrameworkElement;
             THROW_IF_FAILED(buttonIcon.As(&buttonIconAsFrameworkElement));
+            ComPtr<IUIElement> separator;
+            if (iconPlacement == ABI::AdaptiveNamespace::IconPlacement::AboveTitle && allActionsHaveIcons)
+            {
+                THROW_IF_FAILED(buttonContentsStackPanel->put_Orientation(Orientation::Orientation_Vertical));
 
-            // Just add spacing when the icon must be located at the left of the title
-            if (iconPlacement == ABI::AdaptiveNamespace::IconPlacement::LeftOfTitle)
+                // Set height to iconSize (aspect ratio is automatically maintained)
+                THROW_IF_FAILED(buttonIconAsFrameworkElement->put_Height(iconSize));
```

Comment:
having just one dimension for rectangle seems indicating square; maintaining the aspect ratio is the right approach? shouldn't we explicitly set the height and width to iconsize?